pkg/asset/machines/worker.go

```diff
@@ -405,10 +405,9 @@ func (w *Worker) Generate(ctx context.Context, dependencies asset.Parents) error
 				instanceTypes := awsdefaults.InstanceTypes(installConfig.Config.Platform.AWS.Region, arch, configv1.HighlyAvailableTopologyMode)
 				switch pool.Name {
 				case types.MachinePoolEdgeRoleName:
-					ok := awsSetPreferredInstanceByEdgeZone(ctx, instanceTypes, installConfig.AWS, zones)
-					if !ok {
-						logrus.Warnf("failed to find preferred instance type for one or more zones in the %s pool, using default: %s", pool.Name, instanceTypes[0])
-						mpool.InstanceType = instanceTypes[0]
+					if !awsSetPreferredInstanceByEdgeZone(ctx, instanceTypes, installConfig.AWS, zones) {
+						// Using the default instance type from the non-edge pool often fails.
```

Comment:
This seems to be the main part related to the bug. Is it ok to fail here (i.e. is this the intended behavior)?

Reply:
My proposal is: it seems better to fail here instead of having an undetected node failing to come up with the error as in the PR's description. The non-edge node default might not be compatible with edge zones.

Reply:
LGTM failing when any zone isn't found the respective type.